java-sdk/sdk-checkpoints.asciidoc

```diff
@@ -0,0 +1,88 @@
+.. It’s the second job, so checkpoints are stored in the SDK CrawlDB. The controller will send the checkpoints to the fetcher.
+. The fetcher receives and detects the checkpoints. Then, the fetcher emits candidates and updates the checkpoint. The update may take place at a later point, but the checkout _must_ be updated.
+.. If the checkpoint data matches current data, the fetcher will emit the same checkpoint.
+. The SDK controller will process the candidates and update the checkpoint data in the SDK CrawlDB.
+
+
+=== Stopping a Running Job
+
+When a job is stopped, the current state of the job is stored so that it can be completed when the job is resumed.
+
+==== Stop Handling Design
+
+The SDK controller will keep track of the incomplete/complete items during a job. An incomplete item is an item that was emitted as a candidate but has not been sent to the fetcher to be processed. Alternatively, the fetch implementation may not have emitted the FetchResult message for that item. The incomplete item is stored in the SDK CrawlDB and marked as incomplete.
+
+A complete item is an item that was emitted first as a candidate and sent to the fetcher to be processed. The fetcher completes the process by sending a FetchResult message to indicate the process is finished. This item is then marked as complete by the SDK controllers by setting the `blockID` field in the item metadata to match the blockID of the current job.
+
+The blockId is used to quickly identify items in the CrawlDB which may not have been fully processed (completed). A blockId identifies a series of one or more Jobs. The lifetime of a blockId spans from the start of a crawl to the completion. The SDK controller will generate and use a new blockId when:
+
+* The current job is the first job.
+* The previous job’s state is `FINISHED`.
+
+When a job starts and the previous job did not complete, the job's state is `STOPPED`. In this case, the previous job's blockId is reused. The same blockId will be reused until the crawl successfully completes. The SDK controller will continue checking the CrawlDB for incomplete items, which are identified by having a blockID that doesn't match the previous job blockID. This approach ensures all items within the job are completed before the next job beings, even if the job was stopped multiple times before completion.
+
+When an item is considered a candidate, the item's blockId is not added to, or updated in, the item’s metadata. When the item is processed by the fetcher, and a FetchResult was emitted for that item, the blockId is added to the item metadata and stored in the SDK CrawlDB. The item is then marked as complete and will be processed in the next non-stopped job.
+
+When all items are complete, the SDK will check for checkpoints, as detailed in <<checkpoint-design, Checkpoint Design>>.
+
+IMPORTANT: A job won’t be started if there are incomplete items from the previous job stored in the SDK CrawlDB.
```

Comment:
Hmm, I'm not sure what this means... any thoughts @wsolaligue?

Reply:
@mwmitchell AFAIK the job will resolve incomplete items from the previous job, this way the job can be stopped and then get resumed

Reply:
not sure about this statement too. The job will start for sure, but if there are incomplete items in the crawlDB, then **only** those items will be processed in the current job. It's like: if there are incomplete items in the crawlDB, then let's process just the incomplete items.

Reply:
Please see the latest commit. Does this work for everyone?

Reply:
I think it's fine

Reply:
Thanks guys 👍